File: problems/82A - Double Cola.py
import math
import sys

n = int(sys.stdin.readline())

def end_of_loop(n, olval = None) :
    if n == -1 : return 0
    return 5*(2**n) + (olval if olval is not None else end_of_loop(n - 1))
def length(n) :
    return 5*math.pow(2, n)
def closed_form_series(n) :
    return 5*(2**(n+1)) - 5
CHARS = ("Sheldon", "Leonard", "Penny", "Rajesh", "Howard")

pp = 0
olval = 0
while True :
    # val = closed_form_series(pp)
    val = end_of_loop(pp, olval)
    if val >= n :
        break
    olval = val
    pp +=1
start = olval + 1
l = length(pp)
offset = n - start
split = l / 5
char = CHARS[math.floor(offset / split)]
print(char)

File: problemscpp/82A - Double Cola.cpp
#include <iostream>
#include <cmath>

using namespace std;

const std::string CHARS[5] = {"Sheldon", "Leonard", "Penny", "Rajesh", "Howard"};

int end_of_loop(int n, int oldval) {
    if (n == -1)  return 0;
    return 5*pow(2, n) + oldval;
}
int end_of_loop(int n) {
    if (n == -1)  return 0;
    return 5*pow(2, n) + end_of_loop(n - 1);
}

int length(int n) {
    return 5*pow(2, n);
}

int main() {
    std::string n;
    std::getline(std::cin, n);
    int num = std::stoi(n);
    int pp=0;
    int olval =0;
    int val;
    while (true) {
        val = end_of_loop(pp, olval);
        if (val >= num) {
            break;
        }
        olval = val;
        pp +=1;
    }
    int start = olval + 1;
    int l = length(pp);
    int offset = num - start;
    int split = l / 5;
    std::string character = CHARS[offset / split];
    std::cout << character;
    return 0;
}